Support for AWS SSO + custom attributes to configure SNS platform app

MarcPerezPro opened the pull request:
This commit adds the possibility to pass an AWS SNS client, which can be initialized with AWS SSO credentials. It also makes the code unit testable by providing a client with `stub_responses`. I updated the Unit Tests in [this PR](https://github.com/fastlane-community/fastlane-plugin-aws_sns/pull/8).

We also needed to configure the SNS platform app with various attributes like `EventEndpointCreated`, so I added a parameter for that. The full list of SNS platform application attributes can be found [here](https://docs.aws.amazon.com/sns/latest/api/API_SetPlatformApplicationAttributes.html).

Our use-case :
```ruby
sns_app_arn = aws_sns(
  aws_sns_client: Aws::SNS::Client.new,
  platform: 'FCM', # Firebase Cloud Messaging
  platform_fcm_server_key: platform_fcm_server_key,
  platform_name: platform_name,
  update_if_exists: true, # In case we need to rerun this lane.
  attributes_override: {
    EventEndpointCreated: "arn:aws:sns:#{region}:#{account}:bFanCreateEvents",
    EventEndpointUpdated: "arn:aws:sns:#{region}:#{account}:bFanUpdateEvents",
    EventEndpointDeleted: "arn:aws:sns:#{region}:#{account}:bFanDeleteEvents"
  }
)
```

joshdholtz commented on lib/fastlane/plugin/aws_sns/actions/aws_sns_action.rb:
```diff
@@ -138,27 +149,31 @@ def self.details
-                                       default_value: ENV['AWS_ACCESS_KEY_ID']),
+                                      env_name: "AWS_SNS_ACCESS_KEY",
+                                      description: "AWS Access Key ID",
+                                      optional: false,
```
```suggestion
                                      optional: true,
```

joshdholtz commented on lib/fastlane/plugin/aws_sns/actions/aws_sns_action.rb:
```diff
@@ -138,27 +149,31 @@ def self.details
-                                       default_value: ENV['AWS_SECRET_ACCESS_KEY']),
+                                      env_name: "AWS_SNS_SECRET_ACCESS_KEY",
+                                      description: "AWS Secret Access Key",
+                                      optional: false,
```
```suggestion
                                      optional: true,
```

joshdholtz commented on lib/fastlane/plugin/aws_sns/actions/aws_sns_action.rb:
```diff
@@ -138,27 +149,31 @@ def self.details
           FastlaneCore::ConfigItem.new(key: :region,
                                       env_name: "AWS_SNS_REGION",
                                       description: "AWS Region",
                                       optional: false,
```
```suggestion
                                      optional: true,
```

joshdholtz reviewed (changes requested):
I really like these changes! I think my one question is if `:access_key`, `:secret_access_key`, and `:region` should be optional / conflict with `:client` 🤷‍♂️  Thoughts?

return-main commented:
> I really like these changes! I think my one question is if `:access_key`, `:secret_access_key`, and `:region` should be optional / conflict with `:client` 🤷‍♂️ Thoughts?

Hey Josh,
I am a big fan of your work on Fastlane ❤️

I read the code from [config_item.rb](https://github.com/fastlane/fastlane/blob/master/fastlane_core/lib/fastlane_core/configuration/config_item.rb) and tried to modernize the available_options. I made the `:access_key`, `:secret_access_key`, and `:region` conflict with `:aws_sns_client`.
Let me know if I got something wrong.

Also I'd love to know how you test individual Actions, I tried to update [my smoke tests](https://github.com/return-main/fastlane-plugin-aws_sns/blob/feature/unit_testing_overhaul/spec/aws_sns_action_spec.rb) based on [an official actions spec file](https://github.com/fastlane/fastlane/blob/master/fastlane/spec/actions_specs/carthage_spec.rb). I think I'll get the hang of it with time.

joshdholtz commented:
@return-main Hey! So sorry I missed this response 😱 Are you still using this and interested in getting this merged? 😇 

I haven't personally used this in a while so I don't have a good testing environment setup for it but I'll get something setup if this is still something you think should be merged in! 😊

MarcPerezPro committed:
Support for AWS SSO + custom attributes to configure SNS platform app

This commit adds the possibility to pass an AWS SNS client, which can be initialized with AWS SSO credentials. It also makes the code unit testable by providing a client with stub_responses (I'll make a separate PR for that).
We also needed to configure the SNS platform app with various attributes like EventEndpointCreated, so I added a parameter for that. The full list of SNS platform application attributes can be found here: https://docs.aws.amazon.com/sns/latest/api/API_SetPlatformApplicationAttributes.html .

return-main commented:
Hi @joshdholtz 
No worries, I know you've been busy with your family 😄 

I put this in our production fastlane/Pluginfile in the meantime:
```ruby
# Replace with
# gem 'fastlane-plugin-aws_sns'
# when https://github.com/fastlane-community/fastlane-plugin-aws_sns/pull/7 is merged
gem 'fastlane-plugin-aws_sns', git: 'https://github.com/return-main/fastlane-plugin-aws_sns.git', branch: 'feature/attributes_override'
```

We are still interested in getting it merged, so I retested my code and squashed the commits for https://github.com/fastlane-community/fastlane-plugin-aws_sns/pull/7 and https://github.com/fastlane-community/fastlane-plugin-aws_sns/pull/8
